Version 2.0 Rework

gkjohnson opened the pull request:
- Reorganize API to expose Spice functions, CSpice module, and Timecraft kernel loader helper separately.
- Convert to ES6 modules.
- Add an example page demonstrating using kernels and time conversion functions.
- Enable kernels to be loaded dynamically instead of requiring them to be referenced up front using pre and post install.
- Reorganize folders
- Add msl kernels into the repo

CSpice function wrappers are autogenerated in [this project](https://github.com/gkjohnson/generate-cspice-js-wrapper-functions/blob/master/output/GENERATED_FUNCTIONS.md) and can serve as a starting point for wrapping new functions. In some cases the functions must be modified a bit to work correctly, though. That script can be updated to include autogenerating typescript definitions / jsdoc comments.
- Change `package.json` license field from "Apache 2.0" to "Apache-2.0"

**Removed**
- Functions needed for managing spice "cells" (not sure what they do) but that otherwise seem to be unused. If needed this can be added back once a use case is presented.
- "alias" functions that map other spellings of function names to spice function. Ie "furnsh -> furnish" or "convrt -> convert".
- Sets of ["derived"](https://github.com/NASA-AMMOS/timecraftjs/blob/master/spice.js#L1671) functions to help simplify some time conversions. Perhaps these can be added back in a separate helpers class at a later point.
- Wrapper of the [error action set functions](https://github.com/NASA-AMMOS/timecraftjs/blob/master/spice.js#L1898-L1906) which aren't needed.
- Some [wrappers for the chronos command line utility](https://github.com/NASA-AMMOS/timecraftjs/blob/master/spice.js#L1924-L1936) for performing time conversions which otherwise seem doable with other functions.

**TODO**
- [x] List features / dead code removed from original version
- [x] Remove `_old` variants of spice, timecraft scripts
- [x] Verify package.json is up to date
- [x] Update package-lock.json
- [x] Reorganize kernels into proper file structure
- [x] Make function return statements consistent
- [x] Add issues for the following
    - [x] Add typescript definitions #3 
    - [x] Add example demonstrating metakernel loading #4 
    - [x] Investigate rebuilding CSpice with more code culled, using wasm, or base64 encoded data for binary array to get size down further. #5 
    - [x] Enable Github pages example.
    - [x] Add jest tests for kernel loading #6 
    - [x] Update documentation #7 
    - [x] Consistently comment spice.js functions #8

gkjohnson committed:
Add TODO.md

gkjohnson committed:
small adjustments

gkjohnson committed:
move files to src

gkjohnson committed:
Add git ignore, examples

gkjohnson committed:
Get basic example going

gkjohnson committed:
Get example working

gkjohnson committed:
Updates

gkjohnson committed:
remove unused imports

gkjohnson committed:
eslint --fix

gkjohnson committed:
cspice2 -> let const

gkjohnson committed:
renames

gkjohnson committed:
package update

gkjohnson committed:
editor config addition

gkjohnson committed:
Add int and double size constants

gkjohnson committed:
add example with msl times

gkjohnson committed:
simplify build

gkjohnson committed:
update example

gkjohnson committed:
Delete TODO.md

gkjohnson committed:
Update index.html

gkjohnson committed:
Update index.html

gkjohnson committed:
Add kernels needed for sun pos

gkjohnson committed:
Merge branch 'v2-rework' of github.com:gkjohnson/timecraftjs into v2-rework

gkjohnson committed:
Add function for parsing the metakernel

gkjohnson committed:
Delete spice_old.js

gkjohnson committed:
Delete timecraft_old.js

gkjohnson committed:
Update package.json

gkjohnson committed:
update package lock

gkjohnson committed:
Merge branch 'v2-rework' of github.com:gkjohnson/timecraftjs into v2-rework

# Conflicts:
#	package-lock.json

gkjohnson committed:
package lock update

gkjohnson commented:
Ping @camargo -- take a look at the above list and what's been removed and what you think. I'd rather start with a minimal set of usable functions and expand that set as we see a need rather than carrying everything over from the current Timecraft version.

Also:

- Should we wait to update the docs? Or should I do that here? It might be a relatively big change.
- Some of the functions that were carried over from the previous implementation include some custom logic when returning data that is otherwise output from pointers, such as when [handling errors](https://github.com/gkjohnson/timecraftjs/blob/v2-rework/src/spice.js#L1146-L1151) (meaning the calling app can't handle an error so this should definitely be fixed) or when a [flag is returned indicating a field is found](https://github.com/gkjohnson/timecraftjs/blob/v2-rework/src/spice.js#L59-L63). This is inconsistent with other cases where [all output pointers are just returned as a dictionary if there's more than one](https://github.com/gkjohnson/timecraftjs/blob/v2-rework/src/spice.js#L461). I'm inclined to change all functions to use the same dictionary return pattern for the sake of consistency. And if an object has only one output variable then it just returns that value directly. I don't necessarily know what to do if a function has a return value _and_ output variables but we can handle that when we come across that case.

Thoughts?

Thanks again for taking a look!

camargo commented on src/index.js:
```diff
@@ -0,0 +1,23 @@
+import CSpice from './cspice.js';
```
Can we just `export` these files directly? Instead of `importing` first and then `exporting`? For example:
```js
export * as CSpice from './cspice.js';
export * as Spice from './spice.js';
export {
    chronos,
    loadKernel,
    loadKernelFromBuffer,
    parseMetakernel,
    prepareFileFromBuffer,
    removeFile,
    unloadKernel
} from './timecraft.js';
```

gkjohnson replied:
I don't think that valid in es6 modules is it?

gkjohnson replied:
https://stackoverflow.com/questions/38077164/es6-export-from-import

🤯

I'll try making the change.

gkjohnson replied:
Done!

camargo commented:
Commenting on both of your points:
- I don't think we need to update the docs in this PR. Can you add it to the issues we need to create?
-  I agree we should have a consistent return type for all the spice functions. If a function has an output pointer and a return value, I'd say just return both in an object with a consistent property name for the return value.

camargo reviewed (approved):
The code looks great, and I was able to run it locally successfully. Thanks for doing this!

gkjohnson commented:
> I don't think we need to update the docs in this PR. Can you add it to the issues we need to create?

Sounds good.

> I agree we should have a consistent return type for all the spice functions. If a function has an output pointer and a return value, I'd say just return both in an object with a consistent property name for the return value.

Sounds good -- I'll update the existing functions that don't match the existing pattern and then mark this ready for review. So [this function](https://github.com/gkjohnson/timecraftjs/blob/v2-rework/src/spice.js#L1146-L1151) will return `{ sp2000, errmsg }` and [this one](https://github.com/gkjohnson/timecraftjs/blob/v2-rework/src/spice.js#L59-L63) will return `{ name, found }`.

gkjohnson committed:
Rearrange MSL kernels

gkjohnson committed:
Remove commented out functions, clean up return values

gkjohnson committed:
fix function definition

gkjohnson committed:
Fix exports

gkjohnson commented:
Updated the function returns and removed some commented out functions in e44516f and aad1da3 -- if you could take a look over those to make sure I haven't done something dumb that would be great.

And I adjusted the imports in d3fb0af.

gkjohnson commented on spice.js:
```diff
@@ -1,1944 +0,0 @@
-		*/
-		bodvrd: function(body, item, maxn) {
-			var valuesptr = this.Module._malloc(8*maxn);
-			var dimptr = this.Module._malloc(2);
```
I don't know why this was set to "2" previously (changed to INT_SIZE = 4 now) -- the [docs state dim is an INT](https://naif.jpl.nasa.gov/pub/naif/toolkit_docs/FORTRAN/spicelib/bodvrd.html).
